src/box/applier.cc

```diff
@@ -1290,6 +1346,682 @@ applier_on_rollback(struct trigger *trigger, void *event)
+	ibuf_defragment(ibuf);
+	region_reset(&msg->region);
+	stailq_create(&msg->txs);
+	if (tdata->reader != NULL && !fiber_is_dead(tdata->reader))
+		fiber_wakeup(tdata->reader);
+	if (msg->has_ack) {
+		tdata->has_ack = true;
+		vclock_copy(&tdata->ack_vclock, &msg->ack_vclock);
+		if (tdata->writer != NULL && !fiber_is_dead(tdata->writer)) {
+			fiber_cond_signal(&applier->writer_cond);
+		}
+		msg->has_ack = false;
+	}
+}
+
+/**
+ * Given a pair of input buffers, move the unparsed data from the buffer
+ * departing to tx thread to the buffer remaining available in applier thread.
+ */
+static inline size_t
+move_unparsed(struct ibuf *oldbuf, struct ibuf *newbuf)
+{
+	assert(ibuf_used(newbuf) == 0); /* cannot move on top of other data */
+	size_t unparsed = ibuf_unparsed(oldbuf);
+	if (unparsed > 0) {
+		void *ptr = ibuf_alloc(newbuf, unparsed);
+		if (ptr == NULL) {
+			panic("Applier failed to allocate memory for incoming "
+			      "transactions on ibuf");
+		}
+		memcpy(ptr, oldbuf->xpos, unparsed);
+		oldbuf->wpos -= unparsed;
+	}
+	return unparsed;
+}
+
+/** Get the message not containing any data ready for processing. */
+static struct applier_msg *
+applier_thread_next_msg(struct applier_thread_data *tdata)
+{
+	int cur = tdata->msg_ptr;
+	struct applier_msg *msg = &tdata->msgs[cur];
+	struct ibuf *ibuf = &msg->ibuf;
+	if (ibuf_unparsed(ibuf) == ibuf_used(ibuf)) {
+		/*
+		 * The ibuf doesn't contain any parsed data.
+		 * Use current mesage.
+		 */
+		return msg;
+	}
+	cur = (cur + 1) % 2;
+	msg = &tdata->msgs[cur];
+	struct ibuf *other = ibuf;
+	ibuf = &msg->ibuf;
+	if (ibuf_used(ibuf) == 0) {
+		tdata->msg_ptr = cur;
+		move_unparsed(other, ibuf);
+		return msg;
+	}
+	return NULL;
+}
+
+static void
+applier_thread_push_batch(struct applier_thread_data *tdata,
+			  struct applier_msg *msg, struct stailq *txs)
+{
+	assert(msg != NULL);
+	stailq_concat(&msg->txs, txs);
+
+	cmsg_init(&msg->base, tdata->route);
+	cpipe_push(&tdata->tx_pipe, &msg->base);
+}
+
+/**
+ * Read as much data as possible. Do not yield as long as at least one byte is
+ * available.
+ */
+static ssize_t
+applier_read_ahead(struct applier *applier, struct ibuf *ibuf)
+{
+	return coio_breadn_timeout(&applier->io, ibuf, 1,
+				   replication_disconnect_timeout());
+}
+
+static struct applier_tx *
+applier_parse_tx(struct applier *applier, struct ibuf *ibuf,
+		 struct region *region)
+{
+	const char *data = ibuf->xpos;
+	const char *end = ibuf->wpos;
+	uint64_t tsn = 0;
+	struct applier_tx *tx = region_alloc_object_xc(region,
+						       struct applier_tx);
+	stailq_create(&tx->rows);
+	do {
+		if (data == end)
+			goto not_read;
+		if (mp_typeof(*data) != MP_UINT) {
+			tnt_raise(ClientError, ER_INVALID_MSGPACK,
+				  "packet length");
+		}
+		if (mp_check_uint(data, end) > 0) {
+			goto not_read;
+		}
+		uint64_t len = mp_decode_uint(&data);
+		const char *reqend = data + len;
+		if (reqend > end) {
+			goto not_read;
```

Comment:
Do I read correctly that if you fail to read a whole transaction, you simply discard all parsed data and try to parse it again, once more data is available in the buffer?

Reply:
Yes, that's true. I do so, because I have to parse row->is_commit (or row->tsn) to understand whether I have reached the end of the transaction.

So I don't see any solution other than to parse all xrow_headers until is_commit row is found.
I only parse row headers though. The requests are parsed only once I know I've read a full transaction.

I could in theory keep the parsed data for later use, but then I'd have to check whether ibuf was reallocated between my parsing attempts. And this detection looks rather ugly IMO (as can be seen in previous patches about removing row body copying).

BTW, there was a bug with freeing the whole region (with previously parsed transactions), when I couldn't read the next transaction. I've fixed that.

Reply:
IMO it'd be better to copy data to the region you use for storing row headers, because copying is cheaper then decoding. Besides, it'd be done in a separate thread, i.e. performance wouldn't be an issue. Then we could use one ibuf in applier thread for reading rows and wouldn't need `move_unparsed`.

Reply:
After a verbal discussion I've reworked row allocation and storage. Now there's a single ibuf per applier, which is used to read incoming rows.
All the row bodies (and rows themselves) are saved on a lsregion.

applier_read_tx() is reused in applier_subscribe, which has simplified the code quite a bit.

Now there is a separate lsregion per each applier (even in one thread).
I thought of creating a single lsregion per thread, but this is non-trivial:
imagine there are two appliers: the first one reads a synchro request, assigning it an lsregion id 1, and the second one reads a simple ping, assigning it an lsregion id 2.
The requests are pushed in correct order (first 1, then 2), but processing synchro request 1 includes a yield in tx thread. This means that request 2 will be processed first and will return to applier thread first, making it free the data needed by the syncrho request.

In order to fix that I'd have to introduce some list of "lsregion ids currently in use" and only call lsregion_gc() when a message corresponding to the first entry in list returns to the thread.

I decided not to do that now, but I'll implement that if you like a single lsregion per thread more.